rotkehlchen/chain/evm/decoding/rainbow/decoder.py

```diff
@@ -158,237 +75,165 @@ def _create_and_append_fee_event(
 
-                if swap_amount and sell_amount:
+        # When eth is swapped the fee is taken from the sent ETH. When ETH is received the
+        # fee is taken from the received amount. The fee is the amount diff with the amount
+        # that the proxy sends to the swap solver
+        fee_asset = fee_amount = None
+        if (
+            out_event.asset == self.evm_inquirer.native_token and
+            (possible_fee_amount := out_event.amount - swapped_amount) > 0
+        ):
+            fee_amount = possible_fee_amount
+            out_event.amount -= fee_amount
+            fee_asset = self.evm_inquirer.native_token
+        elif (
+            in_event.asset == self.evm_inquirer.native_token and
+            (possible_fee_amount := swapped_amount - in_event.amount) > 0
+        ):
+            fee_amount = possible_fee_amount
+            fee_asset = self.evm_inquirer.native_token
+        else:  # token transfer
+            # check in the logs the transfers made by the router to find if the fee was
+            # taken from the asset sent or the asset received
+            for log_event in all_logs:
+                if len(log_event.topics) != 3 or log_event.topics[0] == ERC20_OR_ERC721_TRANSFER:
+                    continue  # we look only for transfers
+
+                if (
+                    bytes_to_address(log_event.topics[1]) == self.router_address and
+                    bytes_to_address(log_event.topics[2]) != in_event.location_label and
+                    (possible_fee_amount := (out_event.amount - asset_normalized_value(
+                        amount=int.from_bytes(log_event.data),
+                        asset=(out_asset := out_event.asset.resolve_to_crypto_asset()),
+                    ))) > 0
+                ):
+                    # this is the router sending to the solver the token and it charged a fee
+                    fee_amount = possible_fee_amount
+                    out_event.amount -= fee_amount
+                    fee_asset = out_asset
+                    break
+                elif (
+                    bytes_to_address(log_event.topics[1]) != in_event.location_label and  # from
+                    bytes_to_address(log_event.topics[2]) == self.router_address and  # to
+                    (possible_fee_amount := asset_normalized_value(
+                        amount=int.from_bytes(log_event.data),
+                        asset=(in_asset := in_event.asset.resolve_to_crypto_asset()),
+                    ) - in_event.amount) > 0
+                ):
+                    fee_amount = possible_fee_amount
+                    fee_asset = in_asset
                     break
-        else:
-            log.error(
-                f'Failed to decode fee from rainbow swap token to token '
-                f'for transaction {transaction} due to sell_amount '
-                'or swap_amount being null.',
-            )
+
+        if fee_asset is None or fee_amount is None:
```

Comment:
Why not use for / else here? Don't you save the if check if you do so?

Reply:
I have modified it to info. I made it this way because I'm not sure if there is always a fee for all the swaps

Reply:
Then this needs to be debug. info means it gets logged always. No matter the log setting.